### ISSUE TITLE
Memory leak during shape creation using Open Cascade

### ISSUE THREAD

**Issue by derbosper** [IfcOpenShell/IfcOpenShell#671]
There is a memory leak during the call of the create_shape function in Python using OpenCascade. For 530 elements, the amount of leaked memory is about 340 MB. Here is a script that triggers this leak.

```python
import ifcopenshell
from ifcopenshell.geom import create_shape

def generate_model(building_filename):
    settings = ifcopenshell.geom.settings()
    settings.set(settings.USE_PYTHON_OPENCASCADE, True)
    ifc = ifcopenshell.open(building_filename)
    products = ifc.by_type("IfcProduct")

    for product in products:
        if product.Representation is not None:
            shape = create_shape(settings, product)

if __name__ == "__main__":
    generate_model("Duplex_M_20111024_ROOMS_AND_SPACES.ifc")
```

I used [this](https://portal.nibs.org/files/wl/?id=ourEhlbf4MhcqpAn2pvvDFQpMsJ78uOB) sample file from [here](https://www.nibs.org/page/bsa_commonbimfiles). IfcOpenShell version: 0.6.0.a1, oce version: 0.18.3. 

Do you have any idea of possible solutions for this?

**Comment by aothms** [comment]
Quick question: Do you get similar leaks with `USE_PYTHON_OPENCASCADE, False`? 

There are various things at play here. OCCT, IfcOpenShell, SWIG, PythonOCC all these layers complicate memory management. What PythonOCC version are you using? There are a lot of discussions regarding memory management there as well https://github.com/tpaviot/pythonocc-core/issues/214

In essence create_shape() only returns a string serialization of the shape for use in Python, so the error surface there is minimal, but it could be there are deeper underlying issues in the C++ code base of IfcOpenShell related to this.

**Comment by MrBosper** [comment]
Thanks for the reference. I use PythonOOC v0.18.2 via tpaviots conda package. 

The same thing happens when `USE_PYTHON_OPENCASCADE` is set to `False`. Does this mean that OCC library can be excluded as a potential source of the leak?

**Comment by FreakTheMighty** [comment]
I see the same memory leak without PythonOCC installed. As I call `create_shape` the memory slowly creeps up and is never freed.

**Comment by FreakTheMighty** [comment]
Any thoughts on why I don't see this while using the iterator?

Also, I would use the iterator, but I would like to be able to skip elements. Is there anyway to provide more control over what is being iterated. Or, alternatively, perhaps there is a way of deferring mesh generation so that iteration itself is quick.

**Comment by aothms** [comment]
> Any thoughts on why I don't see this while using the iterator?

Thanks that's exactly a useful hint.

The iterator will after every call to `next()` automatically free the previously returned shape memory.

Perhaps we need to double check how `create_shape()` tells SWIG that it's in charge of freeing the returned object. See for example `open()` being annotated with `%newobject` [0]  so that SWIG calls the C++ destructor when the Python reference is garbage collected.

For `create_shape()` it works a bit differently, because it will be wrapped by the typemap [1]. I thought since that uses `SWIG_POINTER_OWN`, memory will be freed appropriately, but there is also a different typemap above so perhaps double check the wrapper code that SWIG generates and what code paths are followed.

[0] https://github.com/IfcOpenShell/IfcOpenShell/blob/master/src/ifcwrap/IfcParseWrapper.i#L409
[1] https://github.com/IfcOpenShell/IfcOpenShell/blob/master/src/ifcwrap/IfcGeomWrapper.i#L143

> Or, alternatively, perhaps there is a way of deferring mesh generation so that iteration itself is quick.

Yes that should indeed be somehow possible. @Moult also needs that I think for IfcBlender. But it's a pain to implement. This Iterator is an ugly beast in C++ and the multithreaded version even follows a completely different approach.

**Comment by FreakTheMighty** [comment]
@aothms sounds like you have some promising leads. Thank you for looking into it!

**Comment by YaroslavKormushyn** [comment]
Is there a workaround for this?

I'm using `create_shape()` to explore the coordinates for the shape in a `TopExp_Explorer` and a `BRepTools_WireExplorer`. I traverse the IFC hierarchy top-down, first to get shapes grouped by `IfcBuildingStorey`, then convert them all to JSON.

The grouping by storey is important. Is there a simple way to use the iterator in order to traverse all geometry in the file first, then group all of the converted geometry by storey? This seems a little bit of a backwards approach, just to get out of a memory leak.

**Comment by aothms** [comment]
> then group all of the converted geometry by storey

This is basically following the `ContainedInStructure` inverse attribute on the element. It'll need a bit of extra care working around all the corner cases.

> This seems a little bit of a backwards approach

Well, keep in mind that IFC is a complex graph structure. Geometry definitions are likely shared by elements on different stories (e.g windows and doors almost guaranteed). So it also makes sense to use an approach that benefits from that.

**Comment by Moult** [comment]
@YaroslavKormushyn you can do:

```python
settings = ifcopenshell.geom.settings()

# Set only the contexts you are interested in, I assume you want the 3D body
model = ifcopenshell.util.representation.get_context("Model") # Some people invalidly put in just model
body = ifcopenshell.util.representation.get_context("Model", "Body")
facetation = ifcopenshell.util.representation.get_context("Model", "Facetation") # Revit bug accommodation
settings.set_context_ids([c.id() for c in [model, body, facetation] if c])

for storey in ifc.by_type("IfcBuildingStorey"):
    elements = ifcopenshell.util.element.get_decomposition(storey)
    iterator = ifcopenshell.geom.iterator(settings, ifc, multiprocessing.cpu_count(), include=elements)
    if iterator.initialize():
        while True:
            shape = iterator.get()
            element = ifc_file.by_guid(shape.guid)
            [ ... etc etc ... ]
```

**Comment by Moult** [comment]
Or if you want the iterator to just process everything then group later (which leads to efficiency due to shared geom between storeys) you can just not use the `include=` filter and get all the elements by `element = ifc_file.by_guid(shape.guid)` then correlate that back to the results of `ifcopenshell.util.element.get_decomposition(storey)`.

The `get_decomposition` util function handles all the ambiguous cases of the question of "what's in this storey" (e.g do you want things that are in spaces in this storey, or do you want subelements of element assemblies too, etc).

**Comment by YaroslavKormushyn** [comment]
@Moult thank you for the nice code example! However, the `ifcopenshell.util.representation.get_context("Model", "Body")` and `ifcopenshell.util.representation.get_context("Model", "Facetation")` result in an error for me, with `AttributeError: 'str' object has no attribute 'by_type'` error message. 
The function definition looks as follows:
```python
def get_context(ifc_file, context, subcontext=None, target_view=None)
```
I'm using the 0.7.0 version from PyPI of ifcopenshell. Maybe there have been some changes to this API?

**Comment by Moult** [comment]
@YaroslavKormushyn ah you're right do get_context(ifc, ...) it was just a typo, I didn't test out my code :)

**Comment by YaroslavKormushyn** [comment]
@Moult the snippet you shared works great, thank you!
I have another question about the geometry part, or rather how one might traverse it.
I'm using a `TopExp_Explorer` and a `BRepTools_WireExplorer` to traverse the points of an element. I see that `shape` also has the points `verts` and `faces`, which should contain basically the same geometry.

**Comment by YaroslavKormushyn** [comment]
> I see that shape also has the points verts and faces

@Moult I was referring to [this example](https://wiki.osarch.org/index.php?title=IfcOpenShell_code_examples#Geometry_processing) which is using the iterator. Upon examining the shape in the debugger using the snippet above, I didn't see the verts and faces attribute. Is that something I should specify explicitly when initializing the iterator?

**Comment by Moult** [comment]
Yeah, there are two approaches, either by default IfcOpenShell will triangulate everything and give you verts/faces that are pure triangles, or if you don't want the triangulation, you can do [disable triangulation and use brep data](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/docs/geom/settings.md#disable_triangulation) which will give you OpenCASCADE serialized TopoDS_Shapes which you can then traverse through the OCC wrappers like you have. So the difference (apart from speed) is that one is triangulated and one isn't.

For the BlenderBIM Add-on and Blender's OpenGL viz, with mesh based visualisation, we use the triangulation function. For FreeCAD, they use the BRep approach.

**Comment by YaroslavKormushyn** [comment]
@Moult this is critical knowledge, thanks for sharing.

The reason I'm asking this is that I experience memory leaks while using the OCC wrappers, which is a completely separate library, which is not on PyPI at all, so I had to press their conda package to a .whl and that might have caused the issues above. If I were to see the triangulated data in order to simply traverse the geometry similar to the OCC wrappers, or rather if there was some approach to project the 3D meshes onto a plane and copy over the resulting vertices to a shapely polygon, this would save me a lot of time and headache.

My knowledge of what ifcopenshell is capable of in this area is limited, unfortunately.

**Comment by Moult** [comment]
@YaroslavKormushyn sure, you can just get triangles, and if your plane is represented by a matrix where Z in the normal you can do `(matrix.inverted() @ coordinate)[0:2]` and feed it into shapely.

I will do some work in consolidating these spread-out docs into the Sphinx docs.

**Comment by Moult** [comment]
I've just updated the docs, I realised that they were missing many new settings that were not documented. I also improved the formatting slightly and made it clearer what the IfcConvert flags were and default values.

 - https://blenderbim.org/docs-python/ifcopenshell-python/geometry_iterator.html
 - https://blenderbim.org/docs-python/ifcopenshell/geometry_settings.html

Hope that helps.

**Comment by Moult** [comment]
No, the edges are based on the 2D curve or the 3D triangles, it's not a culled simplified 2D polygon, I misunderstood what you were after. If you're after something like a drawing, you might want to use the ifcopenshell.draw module, which is ... a little experimental and definitely not documented in the slightest ;) but can essentially generate 2D drawings.

Basically feed your iterator into a ifcopenshell.geom.serializers.svg serialiser. I'm not sure how much help I can provide in a Github comment, it's a pretty new and rough field.

**Comment by YaroslavKormushyn** [comment]
@Moult it's hard to learn how to use something that has few usage examples in the codebase :) It would be interesting to know what [all of these settings](https://vscode.dev/github.com/IfcOpenShell/IfcOpenShell/blob/0446fca56ad133582a05934e2e317714f267df70/src/blenderbim/blenderbim/bim/module/drawing/operator.py#L379) are affecting, but I'm fine with performing my own projections using faces/edges of the shapes unless it doesn't work correctly (more below).

> The edges give you pairs of indices into the 3D points that were also (subedges of..) explicit edges in the IFC model. So they are not ordered polygons yet, but (I believe) if you connect them end-to-end they will give you the cycles you're after.

@aothms thanks for confirming that!

I had the impression that was the true nature of edges. However, I encountered an IfcSpace with some challenging geometry, which is not read by ifcopenshell correctly. Maybe this is worth creating another issue for, but I'll just start it here and transfer over if necessary.

I attached the verts, faces, and edges of the shape in the zip below. For some reason, the faces turn out to cover only the outer part of the shape, without the floor and ceiling. The edges behave in a similar fashion (that's expected) but are missing one of the arcs at the bottom, while the other arc is present (unexpected).

To visualize the intermediate results I used the WKT output of the result shapes.

I'm using the same code as above from @Moult's snippet and configuration with `ifcopenshell_settings.USE_WORLD_COORDS = True`. What could be the issue here? How may I help in debugging this? I cannot send you the source file for this for obvious reasons, but how could I cut out the specific shape from an existing IFC to another IFC if that's necessary?

[shape.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/9748049/shape.zip)
![3D](https://user-images.githubusercontent.com/11539004/194925024-8a1e8763-700a-4200-9a1e-b4f1b3c34ed6.png)
![2D](https://user-images.githubusercontent.com/11539004/194925029-35188d4b-3aaf-4969-a3df-e267191cf516.png)
![Screenshot 2022-10-10 210439](https://user-images.githubusercontent.com/11539004/194927753-73e0d191-3809-4a3f-ac56-e4d913c506f7.png)
![Screenshot 2022-10-10 210603](https://user-images.githubusercontent.com/11539004/194927762-edbbeea9-a412-452b-8bfe-457158f8c3a2.png)

**Comment by Moult** [comment]
Actually I have come across this before where a complex IfcSpace will not be parsed correctly and will be missing the top and bottom face. It could be a bug in IOS or just an invalid self-intersecting shape.

IfcPatch https://blenderbim.org/docs-python/ifcpatch.html has a recipe (https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/recipes/ExtractElements.py) for extracting elements if you want to extract just that one element to share with us.

Alternatively models@ifcopenshell.org is a private email you can send to just for debugging purposes manned by @aothms.

**Comment by aothms** [comment]
Looks like a self-intersecting footprint. I see something suspicious at least in the lower left corner.

![afbeelding](https://user-images.githubusercontent.com/1096535/195020118-2dc71eaa-24a7-4a37-8f86-78c9a0a00a05.png)

I've seen this happening with spaces coming out of Revit. If isolate the space I can have a look. We do have code to handle self intersections.

**Comment by YaroslavKormushyn** [comment]
Extracted that one space into another file, please take a look.
[output.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/9754241/output.zip)

It would be awesome to have some sort of automated fixing of invalid or self-intersecting geometry, since most of out use cases have these to some degree.

**Comment by Moult** [comment]
@aothms I've also emailed a test model coming from Revit with some broken rooms :)

**Comment by aothms** [comment]
I've committed some fixes.

The self-intersecting top and bottom face gets nicely converted into three separate faces now. Note that it's not a manifold mesh still though as the faces on the sides of the prism are unaltered and still intersect.

![afbeelding](https://user-images.githubusercontent.com/1096535/195531904-690a1584-a64c-4bec-8530-40f9ee53769a.png)

@Moult I've also worked on some exceptions in your model, if you think something is still missing let me know, the file is a bit messy.

**Comment by YaroslavKormushyn** [comment]
@aothms I'd love to see the updated version of this. Would you consider releasing such updates as patch version bumps? It would play nicely with package managers as pip and conda, which would allow publishing fixes more frequently.

**Comment by aothms** [comment]
I think until v1.0 we'll probably have this kind of branch + sha approach, we really don't have the time to do proper release planning. I'll tag another alpha release and ping @Krande to fire up the conda-forge recipe if he has time. https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.7.0a5

**Comment by Moult** [comment]
@YaroslavKormushyn when there is a new IfcOpenBot build, there will be a new pip release once a month.

**Comment by YaroslavKormushyn** [comment]
I'm a bit confused about the versioning still.

**Comment by Moult** [comment]
Step 1) "Once in a while", @aothms manually triggers a build from IfcOpenBot. This usually happens after he's done some interesting things in the C++ world like fixing bugs like these.

Step 2) Once that IfcOpenBot build is ready, I usually do extra testing in the Python world via the BlenderBIM Add-on or the other utilities (IfcClash, IfcDiff, etc). This usually shows that it works fine but on rare occasions it highlights some other bugs since we have a pretty extensive Python test suite. Once I see that it's stable, I update the various Makefiles and documentation to point to this new IfcOpenBot build.

Step 3) The PyPI release happens once a month on a fixed day. It pulls whatever is the latest one I've marked as stable in step 2. That's why it might seem pretty old. I used to release more frequently on PyPI but PyPI has a filesize limit and no API to purge older releases so I decreased the frequency.

**Comment by aothms** [comment]
New build came through https://github.com/IfcOpenBot/IfcOpenShell/commit/d97594df9645b74f469de47edf3b93cc6c4b4539#comments

**Comment by YaroslavKormushyn** [comment]
Made a wheel from the last `0.7.0a6` tag, can confirm the self-intersecting geometry is now tessellated correctly.
However, I'm no longer using OCC for my needs, so cannot confirm improvements over at memory leaks from that side.

**Comment by tobis1441** [comment]
I am also experiencing memory leaks. Following the comments I cannot tell if this was fixed and I need a newer version or if it is because of OCC. I will do further analysis on my problem, but would be happy about someone telling me what i should expect
